enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,107 @@
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+* How CVO can get the list of operators which has the preflight checks implemented?
+* What information CVO needs from the operators to create the preflight jobs?
+* Does CVO already has capability to create jobs for preflight checks?
+* How to trigger execution of preflight checks?
+* How CVO will report when it gets the result from preflight checks?
+* What information CVO needs from operators to create the preflight job?
+* How will preflight checks report back success or failure, and the detailed info for administrators?
+
+## Summary
+
+The goal of over the air update(OTA) feature of OpenShift is to make upgrades seamless and make the upgrade a success almost always.
+To improve the upgrade success we should be able to detect if the new version is compatible with the existing version before running the actual upgrade.
+As OpenShift V4 is composed of many operators, we need to give operators ability to run some checks to find out if the new version is compatible with the existing version and use the result to improve the success rate of upgrades.
+
+## Motivation
+
+CVO managed operators should be able to run preflight checks before the actual upgrade.
+It would help operators to find out if they can move to the specific version safely before doing an actual upgrade.
+Also administrators are empowered to run these preflight checks without doing an upgrade.
+
+
+### Goals
+
+* Administrators will be able to run preflight checks without running the upgrade.
+* Give Operators (CVO managed) ability to add checks to check if they can safely upgrade to a given release.
+* Preflight checks would be idempotent in nature.
+
+### Non-Goals
+
+* This enhancement will not add code required for preflight checks for individual operators.
+* Operators will have to figure out the actual code they would like to run as part of preflight checks.
+* The preflight checks should not be used as a migration mechanism.
+
+## Proposal
+
+Operators need to provide a different bin path for the preflight checks i.e. /bin/preflight.
+The binpath need to exist inside of the operator image.
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+The binpath /bin/preflight must be executable and it can be symbolic link to the operator binary or a script or It can be non existent.
+CVO will not fail if the path does not exist in an operator image but it would assume that preflight checks are not implemented for the respective operator.
+Also there would be a possible way to skip the preflight checks if someone wants to skip it explicitly.
```

Comment:
Is "someone" an operator maintainer, saying "I have no preflight check and am comfortable updating from any possible source cluster", or a cluster admin, saying "I see that preflight A has failed, and wish to continue with the update regardless"?

Reply:
I'm assuming this is the admin overriding the check. CEE folks tell us that they want a blanket method to force things, if that's what they want then `--force` and all the implications therein seem fine to me. But I think we should refrain from documenting that and provide a fine grained method that allows the admin to enumerate the pre-flights they've chosen to override and that list of overrides should be recorded in the history.

Reply:
Yes, this is for admins who wants to override the check. My prior experience in preflight checks taught me that there will be scenarios when preflight checks might fail because of some peculiar environments. The issue would be in the customer's setup and not in the preflight code.  So having a way to skip it will be helpful. Also it will be helpful for developers who are testing some code changes and they are not going to run preflights during devlopment to reduce the time for upgrade.

Reply:
I'm fine with `--force` undocumented, unsupported safety valve.  I think preflights that fail in valid environments are bugs that don't deserve elegant workarounds.